### PR TITLE
support: invalid cache through rpc

### PR DESCRIPTION
**Pull request by javeme**
Change-Id: I3fe18cd8c893a84a56fb66df651b78c5e0fe5f4b

### CODE REVIEW

**Comment by javeme** [comment]
Cassandra/Scylla ci error:
image

HBase ci error:
image

File: hugegraph-api/src/main/java/com/baidu/hugegraph/rpc/RpcServer.java [hunk 1]
```diff
@@ -24,39 +24,28 @@
...
         this.conf = conf;
-        this.serverConfig = new ServerConfig()
-                            .setProtocol(conf.get(ServerOptions.RPC_PROTOCOL))
-                            .setPort(conf.get(ServerOptions.RPC_SERVER_PORT))
-                            .setHost(conf.get(ServerOptions.RPC_SERVER_HOST))
-                            .setDaemon(false);
+        this.serverConfig = new ServerConfig();
+        this.serverConfig.setProtocol(conf.get(ServerOptions.RPC_PROTOCOL))
+                         .setPort(conf.get(ServerOptions.RPC_SERVER_PORT))
+                         .setHost(conf.get(ServerOptions.RPC_SERVER_HOST))
```
Comment (Linary) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/rpc/RpcServer.java:47]: Let host before port?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CacheNotifier.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+package com.baidu.hugegraph.backend.cache;
+
+import com.baidu.hugegraph.backend.id.Id;
+import com.baidu.hugegraph.type.HugeType;
+
+public interface CacheNotifier extends AutoCloseable {
+
+    public void invalid(HugeType type, Id id);
+
+    public void invalid2(HugeType type, Object[] ids);
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CacheNotifier.java:29]: any friendship name?

**Review by Linary — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ConfigAuthenticator.java [hunk 3]
```diff
@@ -45,8 +45,8 @@ public ConfigAuthenticator() {
 
     @Override
     public void setup(HugeConfig config) {
-        this.tokens.put(USER_ADMIN, config.get(ServerOptions.ADMIN_TOKEN));
-        this.tokens.putAll(config.getMap(ServerOptions.USER_TOKENS));
+        this.tokens.put(USER_ADMIN, config.get(ServerOptions.AUTH_ADMIN_TOKEN));
+        this.tokens.putAll(config.getMap(ServerOptions.AUTH_USER_TOKENS));
```
[the remark this answers is not in the document]
Comment (javeme) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ConfigAuthenticator.java:49]: we can change their order to ensure the admin stays in the map

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CacheNotifier.java (hunk 2 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CacheNotifier.java:29]: sofa-rpc don't allowed to use the same method name in a service, may be names like invalidBatch() is more friendship, here just to keep the name short.

**Review by Linary — APPROVED** [review]

**Review by Linary — DISMISSED** [review]

**Review by zhoney — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/rpc/RpcConsumerConfig.java [hunk 4]
```diff
@@ -19,43 +19,172 @@
...
+            if (responses.size() > 0) {
+                /*
+                 * Just choose the first one as result to return, ignore others
+                 * TODO: maybe more strategies should be provided
+                 */
+                return responses.get(0);
+            } else if (excepts.size() > 0) {
+                throw excepts.get(0);
+            } else {
+                assert providers.isEmpty();
```
[the remark this answers is not in the document]
Comment (javeme) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/rpc/RpcConsumerConfig.java:157]: empty is ok

**Review by Linary — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**